Tools/Scripts/libraries/webkitapipy/webkitapipy/macho.py

```diff
@@ -0,0 +1,128 @@
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import subprocess
+import re
+from pathlib import Path
+from typing import Iterator, NamedTuple
+
+method_to_selector = re.compile(r'[-+]\[(?P<class>\S+) (?P<selector>[^\]]+)\]')
+
+
+class Report(NamedTuple):
+    exports: list[str]
+    selectors: set[str]
+    bindings: list[str]
+    selrefs: list[str]
+    file: Path
+
+
+def report(binary_path: Path, arch: str, symbols_only=False) -> Report:
+    # Symbol table
+    nm_lines = subprocess.check_output(
+        ('xcrun', 'nm', '-P', '--arch', arch, binary_path),
+        text=True, errors='replace'
+    ).splitlines()
+
+    exports = []
+    selectors = set()
+    bindings = []
+    selrefs = []
+
+    for line in nm_lines:
+        if not line or f'for architecture {arch}' in line:
+            # nm always
+            continue
+        symbol, flag, *_ = line.rsplit(maxsplit=4)
+        if flag in 'ATDBCS':
+            exports.append(symbol)
+        elif flag == 't':
+            m = method_to_selector.match(symbol)
+            if m:
+                selectors.add(m.group('selector'))
+        elif flag == 'U':
+            bindings.append(symbol)
+
+    if not symbols_only:
+
+        # Objective-C metadata
+        otool_lines = subprocess.check_output(
+            ('xcrun', 'otool', '-ovV', '-arch', arch, binary_path),
+            text=True, errors='replace'
+        ).splitlines()
+
+        it = iter(otool_lines)
+        for line in it:
+            if line == 'Contents of (__DATA_CONST,__objc_selrefs) section' or \
+                    line == 'Contents of (__DATA,__objc_selrefs) section':
+                for line in it:
+                    if line.startswith('Contents of'):
+                        break
+                    _, name = line.split(maxsplit=1)
+                    selrefs.append(name)
+                break
+
+        # <wtf/SoftLinking.h> dlsym metadata
+        otool_lines = subprocess.check_output(
+            ('xcrun', 'otool', '-vV', '-s', '__TEXT', '__dlsym_cstr',
+             '-arch', arch, binary_path),
+            text=True, errors='replace'
+        ).splitlines()
+
+        it = iter(otool_lines)
+        for name in _otool_parse_string_section(it, '__TEXT,__dlsym_cstr'):
+            bindings.append(f'_{name}')
+
+        otool_lines = subprocess.check_output(
+            ('xcrun', 'otool', '-vV', '-s', '__TEXT', '__getClass_cstr',
+             '-arch', arch, binary_path),
+            text=True, errors='replace'
+        ).splitlines()
+
+        it = iter(otool_lines)
```

Comment:
Really don't like continued reassignments to `it` and `otool_lines`. I think `iter(otool_lines)` could be inlined in most places `it` is being passed rather than used directly.